### PR TITLE
Prove poly decompress

### PR DESCRIPTION
**Pull request by hanno-becker**
Re-opening #161 on the main repo to run CI.

### CODE REVIEW

**Comment by rod-chapman** [comment]
Root cause: cbmc/proofs/poly_decompress/Makefile needs to change to say "--smt2" instead of "--bitwuzla".

The NIX config does build or install bitwuzla, but Z3 (which is used when --smt2 is specified) works OK on this proof.

**Comment by hanno-becker** [comment]
@rod-chapman I added bitwuzla to the cbmc config for now to avoid such local vs. CI mismatches in the future.

File: cbmc/default.nix [hunk 1]
```diff
@@ -21,5 +22,6 @@ builtins.attrValues {
   cbmc-viewer = callPackage ./cbmc-viewer.nix { }; # 3.9
 
   inherit
-    z3_4_12; # 4.12.5
+    z3_4_12# 4.12.5
+    bitwuzla;
```
Comment (mkannwischer) [inline cbmc/default.nix:26]:
```suggestion
    bitwuzla; # 0.4.0
```

File: mlkem/poly.c [hunk 2]
```diff
@@ -220,36 +220,62 @@ void poly_compress(uint8_t r[KYBER_POLYCOMPRESSEDBYTES], const poly *a)
...
+            __CPROVER_loop_invariant(j >= 0)
+            __CPROVER_loop_invariant(j <= 8)
+            __CPROVER_loop_invariant(i >= 0)
+            __CPROVER_loop_invariant(i <= num_blocks)
+            __CPROVER_loop_invariant(__CPROVER_forall { int k; k >= 0 && k <= (8 * i + j - 1) ==>
+                                     ( r->coeffs[k] >= 0 && r->coeffs[k] < KYBER_Q )})
+            __CPROVER_decreases(8 - j)
         {
-            // REF-CHANGE: Truncation happened before
             r->coeffs[8 * i + j] = ((uint32_t) t[j] * KYBER_Q + 16) >> 5;
```
Comment (mkannwischer) [inline mlkem/poly.c:279]: It's rather ugly that for the `(KYBER_POLYCOMPRESSEDBYTES == 128)` case this computation is moved to a separate function, while for the ` (KYBER_POLYCOMPRESSEDBYTES == 160)` case, it's still inlined. 

I don't understand why it needs to be moved into a separate function at all. Can we keep it inlined for both?

**Review by mkannwischer — APPROVED** [review]
Small nits. Rest is looking good to me and CBMC proofs pass on my machine. 

I get a lot of warnings for CBMC, but that seems unrelated to this PR.
```
WARNING: Skipping malformed coverage data in /home/mjk/git/mlkem-c-aarch64/cbmc/proofs/poly_decompress/logs/coverage.xml.

WARNING: Use the --verbose to see what coverage data was skipped.

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_poly_compress

WARNING:   Old symbol pqcrystals_kyber512_ref_poly_compress: file mlkem/poly.h, line 72

WARNING:   New symbol pqcrystals_kyber512_ref_poly_compress: file mlkem/poly.c, line 139

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_poly_decompress

WARNING:   Old symbol pqcrystals_kyber512_ref_poly_decompress: file mlkem/poly.h, line 84

WARNING:   New symbol pqcrystals_kyber512_ref_poly_decompress: file mlkem/poly.c, line 221

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_poly_decompress

WARNING:   Old symbol pqcrystals_kyber512_ref_poly_decompress: file mlkem/poly.h, line 84

WARNING:   New symbol pqcrystals_kyber512_ref_poly_decompress: file MISSING, line 0

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_compress_q_16

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_compress_q_16: file mlkem/poly.h, line 36

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_compress_q_16: file mlkem/poly.c, line 24

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_compress_q_32

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_compress_q_32: file mlkem/poly.h, line 49

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_compress_q_32: file mlkem/poly.c, line 66

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_decompress_q_16

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_decompress_q_16: file mlkem/poly.h, line 43

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_decompress_q_16: file mlkem/poly.c, line 53

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_decompress_q_32

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_decompress_q_32: file mlkem/poly.h, line 56

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_decompress_q_32: file mlkem/poly.c, line 95

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_signed_to_unsigned_q_16

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_signed_to_unsigned_q_16: file mlkem/poly.h, line 62

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_signed_to_unsigned_q_16: file mlkem/poly.c, line 118

WARNING: Skipping source file annotation: wrapped functions for code contracts

```

File: mlkem/poly.c (hunk 2 above)
Comment (hanno-becker) [reply mlkem/poly.c]: Good catch, it should be one or the other. I vaguely recall that proof development was simpler when hoisting out the scalar functions, but apparently everything still works when inlining their bodies, as witnessed by #165 -- so either way will be OK.

For this PR, however, I'd stick with the present approach of _having_ wrappers for scalar functions, and use them consistently. I added a commit using `scalar_decompress_

Comment (mkannwischer) [reply mlkem/poly.c]: Okay, I'm fine with consistently using the scalar functions for now. Thanks.

**Comment by hanno-becker** [comment]
```
WARNING: Skipping malformed coverage data in /home/mjk/git/mlkem-c-aarch64/cbmc/proofs/poly_decompress/logs/coverage.xml.

WARNING: Use the --verbose to see what coverage data was skipped.

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_poly_compress

WARNING:   Old symbol pqcrystals_kyber512_ref_poly_compress: file mlkem/poly.h, line 72

WARNING:   New symbol pqcrystals_kyber512_ref_poly_compress: file mlkem/poly.c, line 139

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_poly_decompress

WARNING:   Old symbol pqcrystals_kyber512_ref_poly_decompress: file mlkem/poly.h, line 84

WARNING:   New symbol pqcrystals_kyber512_ref_poly_decompress: file mlkem/poly.c, line 221

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_poly_decompress

WARNING:   Old symbol pqcrystals_kyber512_ref_poly_decompress: file mlkem/poly.h, line 84

WARNING:   New symbol pqcrystals_kyber512_ref_poly_decompress: file MISSING, line 0

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_compress_q_16

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_compress_q_16: file mlkem/poly.h, line 36

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_compress_q_16: file mlkem/poly.c, line 24

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_compress_q_32

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_compress_q_32: file mlkem/poly.h, line 49

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_compress_q_32: file mlkem/poly.c, line 66

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_decompress_q_16

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_decompress_q_16: file mlkem/poly.h, line 43

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_decompress_q_16: file mlkem/poly.c, line 53

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_decompress_q_32

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_decompress_q_32: file mlkem/poly.h, line 56

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_decompress_q_32: file mlkem/poly.c, line 95

WARNING: Skipping redefinition of symbol name: pqcrystals_kyber512_ref_scalar_signed_to_unsigned_q_16

WARNING:   Old symbol pqcrystals_kyber512_ref_scalar_signed_to_unsigned_q_16: file mlkem/poly.h, line 62

WARNING:   New symbol pqcrystals_kyber512_ref_scalar_signed_to_unsigned_q_16: file mlkem/poly.c, line 118

WARNING: Skipping source file annotation: wrapped functions for code contracts
```

@rod-chapman Do you understand the cause for all those warnings? I get them, too, but usually use ignore them.

**Pull request merged**